pkg/csi/service/node.go

```diff
@@ -680,109 +690,272 @@ func (driver *vsphereCSIDriver) NodeGetInfo(
+	}
+
+	// Fetch node object to set owner ref.
+	nodeObj, err := getNodeObject(ctx, nodeName)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+	// Create spec for CSINodeTopology.
+	csiNodeTopologySpec := &csinodetopologyv1alpha1.CSINodeTopology{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nodeName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "v1",
+					Kind:       "Node",
+					Name:       nodeObj.Name,
+					UID:        nodeObj.UID,
+				},
+			},
+		},
+		Spec: csinodetopologyv1alpha1.CSINodeTopologySpec{
+			NodeID: nodeID,
+		},
+	}
+	// Create CSINodeTopology CR for the node.
+	err = controllerRuntimeClient.Create(ctx, csiNodeTopologySpec)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
 			return nil, logger.LogNewErrorCodef(log, codes.Internal,
-				"failed to get VirtualCenterConfig from cns config. err: %v", err)
+				"failed to create CSINodeTopology CR. Error: %+v", err)
 		}
-		vcManager := cnsvsphere.GetVirtualCenterManager(ctx)
-		vcenter, err := vcManager.RegisterVirtualCenter(ctx, vcenterconfig)
-		if err != nil {
-			return nil, logger.LogNewErrorCodef(log, codes.Internal,
-				"failed to register vcenter with virtualCenterManager. err: %v", err)
+	} else {
+		log.Infof("Successfully created a CSINodeTopology instance for NodeName: %q", nodeName)
+	}
+
+	// Watch on the CSINodeTopology CR.
+	csiNodeTopologyWatcher, err := k8s.NewCSINodeTopologyWatcher(ctx, config)
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to create a watcher for CSINodeTopology CR. Error: %+v", err)
+	}
+	timeoutSeconds := int64((time.Duration(getCSINodeTopologyWatchTimeoutInMin(ctx)) * time.Minute).Seconds())
+	watchCSINodeTopology, err := csiNodeTopologyWatcher.Watch(metav1.ListOptions{
+		FieldSelector:  fields.SelectorFromSet(fields.Set{"metadata.name": nodeID}).String(),
+		TimeoutSeconds: &timeoutSeconds,
+		Watch:          true,
+	})
+	if err != nil {
+		return nil, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to watch on CSINodeTopology instance with name %q. Error: %+v", nodeName, err)
+	}
+	defer watchCSINodeTopology.Stop()
+
+	// Check if status gets updated in the instance within the given timeout seconds.
+	for event := range watchCSINodeTopology.ResultChan() {
+		csiNodeTopologyInstance, ok := event.Object.(*csinodetopologyv1alpha1.CSINodeTopology)
+		if !ok {
+			log.Debugf("Received unidentified object - %+v", event.Object)
+			continue
 		}
-		defer func() {
-			if vcManager != nil {
-				err = vcManager.UnregisterAllVirtualCenters(ctx)
-				if err != nil {
-					log.Errorf("UnregisterAllVirtualCenters failed. err: %v", err)
+		switch csiNodeTopologyInstance.Status.Status {
+		case csinodetopologyv1alpha1.CSINodeTopologySuccess:
+			accessibleTopology := make(map[string]string)
+			topologyLabels := csiNodeTopologyInstance.Status.TopologyLabels
+			if len(topologyLabels) > 0 {
```

Comment:
we do need a len check for `for..range` loop

Reply:
Removed it